File: server/tinytalk/protocol.py
"""Wire protocol between the phone client and this server.

Control messages are JSON text frames. Audio travels as binary frames and is
not represented here: mic audio in and TTS audio out are both 24 kHz mono
PCM16 LE -- one consistent sample rate, matching what Kyutai's STT model and
Mimi audio codec natively expect, so nothing needs resampling.

Ordering requirement: the server only accepts mic audio frames while it is
LISTENING, which it enters on `speech_start` or `interrupt`. A client must
therefore send the control frame (`speech_start` for a fresh utterance,
`interrupt` for a barge-in) *before* any audio frames for that utterance.
Audio that arrives outside LISTENING (e.g. a frame sent before the control
frame, or one still in flight when an utterance ended) is silently dropped —
no error is surfaced. Getting this ordering wrong on a barge-in would lose
the first word of what the child says.

turn_id: the client assigns a new turn_id every time it starts a fresh
listening turn (speech_start or interrupt) and echoes it on nothing else --
the server just remembers whichever turn_id it was most recently told and
stamps every one of its own events (transcript_partial/transcript_final/
response_text/turn_end/error) with that value, until superseded by the next
speech_start/interrupt. This exists because the server's read loop is
strictly serial and STT can take several real seconds per utterance: if the
child interrupts and starts a new utterance while the server is still
finishing the previous one, that stale reply arrives *after* the client has
already moved on to a newer turn. Without turn_id, the client has no way to
tell a late reply for an old, already-abandoned utterance apart from a
legitimate reply for its current one -- confirmed on real hardware to cause
a reply being silently misattributed to the wrong turn, or dropped
entirely, when the child spoke faster than the server could keep up.

`object_seen` is deliberately exempt from both the turn_id contract above
and the speech_start/interrupt ordering requirement -- see the ObjectSeen
dataclass's own docstring for why.
"""

from __future__ import annotations

import json
from dataclasses import dataclass


class ProtocolError(ValueError):
    """Raised when an incoming control message is malformed or unknown."""


LLM_BACKENDS = ("ollama", "groq")


@dataclass(frozen=True)
class SpeechStart:
    """The client's VAD detected speech onset; audio frames follow."""

    turn_id: int


@dataclass(frozen=True)
class SpeechEnd:
    """The client's VAD detected speech offset; finalize the transcript."""


@dataclass(frozen=True)
class Interrupt:
    """The child barged in. Abort the in-flight turn and start listening."""

    turn_id: int


@dataclass(frozen=True)
class ObjectSeen:
    """The child took a photo and on-device Vision classified it; label is
    the recognized object's plain-English name (e.g. "teddy bear").
    Deliberately carries no turn_id, unlike SpeechStart/Interrupt -- taking
    a photo isn't tied to a specific turn boundary, it's queued and woven
    into whichever turn happens next. See
    docs/superpowers/specs/2026-08-29-object-recognition-design.md."""

    label: str


@dataclass(frozen=True)
class NewStory:
    """The client wants to abandon the current story and start a fresh
    one, without tearing down the connection or session -- see
    SessionRunner.handle_new_story(). No turn_id: unlike speech_start/
    interrupt, this isn't itself the start of a turn, so there is nothing
    for it to be echoed back against."""


@dataclass(frozen=True)
class ListStories:
    """Request the saved-story list for the Library screen. No turn_id --
    browsing saved stories is unrelated to live turn-taking."""


@dataclass(frozen=True)
class GetStory:
    """Request one saved story's rewritten pages for the Reading
    screen."""

    story_id: str


@dataclass(frozen=True)
class SynthesizePage:
    """Request on-demand TTS audio for one page of a saved story (the
    Reading screen's per-page replay button). Audio streams through the
    same binary-frame pathway as live TTS, followed by a
    page_audio_done marker."""

    story_id: str
    page_index: int


@dataclass(frozen=True)
class GetPageImage:
    """Request the generated illustration for one page of a saved story
    (the Reading screen's page art). Image bytes stream through the same
    binary-frame pathway as SynthesizePage's audio, followed by a
    page_image_done marker whose has_image field tells the client
    whether a binary frame was actually sent -- a page with no
    illustration (not yet generated, or dropped by the safety check)
    sends the marker only, no binary frame."""

    story_id: str
    page_index: int


@dataclass(frozen=True)
class ConcludeStory:
    """The child (or parent) asked to finish the current story right now
    (the design's "Finish this story" menu item). Carries a turn_id like
    SpeechStart/Interrupt: it results in one more real
    response_text/turn_end pair the client must be able to attribute to a
    turn."""

    turn_id: int


@dataclass(frozen=True)
class SyncDemoStories:
    """The phone hands over stories completed away from home (see
    docs/superpowers/specs/2026-09-09-away-from-home-demo-mode-design.md)
    once it reconnects to the home server. No turn_id -- this isn't part
    of live turn-taking, same reasoning as ListStories/GetStory."""

    stories: tuple[dict, ...]


@dataclass(frozen=True)
class UpdateSettings:
    """Parent-adjustable settings from the Settings screen -- persisted
    client-side, sent once after connecting and again whenever changed
    while connected. Applied to the next story construction, not
    retroactively to one already in progress -- see SessionRunner's
    handle_update_settings().

    llm_backend is optional (docs/superpowers/specs/
    2026-09-22-server-llm-backend-toggle-design.md): absent means "keep
    whatever the server is already using", so an older phone build or
    demo mode's own messages change nothing."""

    target_turns: int
    page_count: int
    llm_backend: str | None = None


ClientMessage = (
    SpeechStart
    | SpeechEnd
    | Interrupt
    | ObjectSeen
    | NewStory
    | ListStories
    | GetStory
    | SynthesizePage
    | GetPageImage
    | ConcludeStory
    | SyncDemoStories
    | UpdateSettings
)

_CLIENT_MESSAGE_TYPES: dict[str, type] = {
    "speech_start": SpeechStart,
    "speech_end": SpeechEnd,
    "interrupt": Interrupt,
    "object_seen": ObjectSeen,
    "new_story": NewStory,
    "list_stories": ListStories,
    "get_story": GetStory,
    "synthesize_page": SynthesizePage,
    "get_page_image": GetPageImage,
    "conclude_story": ConcludeStory,
    "sync_demo_stories": SyncDemoStories,
    "update_settings": UpdateSettings,
}
_TYPES_REQUIRING_TURN_ID = (SpeechStart, Interrupt, ConcludeStory)


def decode_client_message(raw: str) -> ClientMessage:
    try:
        payload = json.loads(raw)
    except json.JSONDecodeError as exc:
        raise ProtocolError(f"control frame is not valid JSON: {raw!r}") from exc

    if not isinstance(payload, dict):
        raise ProtocolError(
            f"control frame must be a JSON object, got {type(payload).__name__}"
        )

    kind = payload.get("type")
    message_type = _CLIENT_MESSAGE_TYPES.get(kind)
    if message_type is None:
        raise ProtocolError(f"unknown client message type: {kind!r}")
    if message_type in _TYPES_REQUIRING_TURN_ID:
        turn_id = payload.get("turn_id")
        if not isinstance(turn_id, int):
            raise ProtocolError(f"{kind} requires an integer turn_id: {raw!r}")
        return message_type(turn_id=turn_id)
    if message_type is ObjectSeen:
        label = payload.get("label")
        if not isinstance(label, str) or not label.strip():
            raise ProtocolError(f"object_seen requires a non-empty string label: {raw!r}")
        return ObjectSeen(label=label.strip())
    if message_type is GetStory:
        story_id = payload.get("story_id")
        if not isinstance(story_id, str) or not story_id.strip():
            raise ProtocolError(f"get_story requires a non-empty string story_id: {raw!r}")
        return GetStory(story_id=story_id.strip())
    if message_type is SynthesizePage:
        story_id = payload.get("story_id")
        page_index = payload.get("page_index")
        if not isinstance(story_id, str) or not story_id.strip():
            raise ProtocolError(
                f"synthesize_page requires a non-empty string story_id: {raw!r}"
            )
        if not isinstance(page_index, int):
            raise ProtocolError(f"synthesize_page requires an integer page_index: {raw!r}")
        return SynthesizePage(story_id=story_id.strip(), page_index=page_index)
    if message_type is GetPageImage:
        story_id = payload.get("story_id")
        page_index = payload.get("page_index")
        if not isinstance(story_id, str) or not story_id.strip():
            raise ProtocolError(
                f"get_page_image requires a non-empty string story_id: {raw!r}"
            )
        if not isinstance(page_index, int):
            raise ProtocolError(f"get_page_image requires an integer page_index: {raw!r}")
        return GetPageImage(story_id=story_id.strip(), page_index=page_index)
    if message_type is SyncDemoStories:
        stories = payload.get("stories")
        if not isinstance(stories, list) or not all(isinstance(s, dict) for s in stories):
            raise ProtocolError(f"sync_demo_stories requires a list of story objects: {raw!r}")
        return SyncDemoStories(stories=tuple(stories))
    if message_type is UpdateSettings:
        target_turns = payload.get("target_turns")
        page_count = payload.get("page_count")
        if not isinstance(target_turns, int):
            raise ProtocolError(
                f"update_settings requires an integer target_turns: {raw!r}"
            )
        if not isinstance(page_count, int):
            raise ProtocolError(
                f"update_settings requires an integer page_count: {raw!r}"
            )
        llm_backend = payload.get("llm_backend")
        if llm_backend is not None and llm_backend not in LLM_BACKENDS:
            raise ProtocolError(
                f"update_settings llm_backend must be one of {LLM_BACKENDS}: {raw!r}"
            )
        return UpdateSettings(
            target_turns=target_turns, page_count=page_count, llm_backend=llm_backend
        )
    return message_type()


def encode_transcript_partial(text: str, turn_id: int) -> str:
    return json.dumps({"type": "transcript_partial", "text": text, "turn_id": turn_id})


def encode_transcript_final(text: str, turn_id: int) -> str:
    return json.dumps({"type": "transcript_final", "text": text, "turn_id": turn_id})


def encode_response_text(text: str, turn_id: int) -> str:
    return json.dumps({"type": "response_text", "text": text, "turn_id": turn_id})


def encode_turn_end(turn_id: int) -> str:
    return json.dumps({"type": "turn_end", "turn_id": turn_id})


def encode_error(message: str, turn_id: int) -> str:
    return json.dumps({"type": "error", "message": message, "turn_id": turn_id})


def encode_arc_stage(stage: str, turn_id: int) -> str:
    return json.dumps({"type": "arc_stage", "stage": stage, "turn_id": turn_id})


def encode_story_list(stories: list[dict]) -> str:
    return json.dumps({"type": "story_list", "stories": stories})


def encode_story_detail(story: dict) -> str:
    return json.dumps({"type": "story_detail", **story})


def encode_page_audio_done(story_id: str, page_index: int) -> str:
    return json.dumps(
        {"type": "page_audio_done", "story_id": story_id, "page_index": page_index}
    )


def encode_page_image_done(story_id: str, page_index: int, has_image: bool) -> str:
    return json.dumps(
        {
            "type": "page_image_done",
            "story_id": story_id,
            "page_index": page_index,
            "has_image": has_image,
        }
    )


def encode_rewriting_started(
    story_id: str | None = None, epilogue: str | None = None
) -> str:
    """`story_id` names the story being rewritten; `epilogue` is its
    fact line (storybook.early_epilogue()), sent now so The End can show
    it without waiting on the whole rewrite (issue #77). Each key is
    omitted when None, so an older client sees the same bare message."""
    payload: dict = {"type": "rewriting_started"}
    if story_id is not None:
        payload["story_id"] = story_id
    if epilogue is not None:
        payload["epilogue"] = epilogue
    return json.dumps(payload)


def encode_rewriting_done() -> str:
    return json.dumps({"type": "rewriting_done"})


def encode_llm_backend(requested: str, active: str, groq_available: bool) -> str:
    """Reply to an update_settings carrying llm_backend -- `active` is what
    the NEXT story will actually use (ollama if groq was requested but this
    server has no GROQ_API_KEY). No turn_id: not part of live turn-taking,
    same as story_list."""
    return json.dumps(
        {
            "type": "llm_backend",
            "requested": requested,
            "active": active,
            "groq_available": groq_available,
        }
    )

File: ios/TinyTalkCore/Sources/TinyTalkCore/Protocol.swift
/// Wire protocol between this client and the Mac server. Mirrors
/// `server/tinytalk/protocol.py` exactly -- same type strings, same field
/// names. Control messages are JSON text frames; audio travels as binary
/// frames and is not represented here (see ServerConnecting in
/// Interfaces.swift for how binary frames are handled).
///
/// turn_id: see protocol.py's module docstring for the full rationale --
/// confirmed necessary on real hardware, not speculative. This client
/// assigns a new turn_id every time it starts a fresh listening turn
/// (speechStart or interrupt) and sends it to the server; the server
/// stamps every event it sends back with whichever turn_id it was most
/// recently told. SessionCoordinator uses this to discard a reply that
/// arrives for a turn the client has already abandoned, rather than
/// misattributing it to whatever turn happens to be active when it
/// arrives (or silently losing it) -- the mechanism that was missing when
/// the server's serialized, multi-second-per-utterance STT pipeline fell
/// behind a child talking faster than it could keep up.
import Foundation

public enum ClientMessage: Sendable, Equatable {
    case speechStart(turnId: Int)
    case speechEnd
    case interrupt(turnId: Int)
    /// See object_recognition.py / this file's `ClientMessage` mirror --
    /// deliberately no turn_id, matching protocol.py's ObjectSeen.
    case objectSeen(label: String)
    /// Abandon the current story and start fresh, without tearing down
    /// the connection -- see server/tinytalk/session.py's
    /// handle_new_story(). No turn_id: mirrors protocol.py's NewStory,
    /// which isn't itself the start of a turn.
    case newStory
    /// Hands one or more stories completed away from home (Groq-backed
    /// demo mode, no persistent server session) to the real server once
    /// reconnected -- see PendingDemoStore/AppModel.connect(). Uses
    /// JSONSerialization in encode() below, not hand-built interpolation
    /// like every other case here: this is the one payload carrying
    /// arbitrary user-generated transcript/reply text.
    case syncDemoStories(stories: [PendingDemoStoryPayload])
    /// Request the saved-story list for the Library screen -- see
    /// server/tinytalk/protocol.py's ListStories. No turn_id: browsing
    /// saved stories is unrelated to live turn-taking.
    case listStories
    /// Request one saved story's full detail (title/pages/epilogue/
    /// rewrite_status) for the Reading/The End screens -- see
    /// protocol.py's GetStory.
    case getStory(storyId: String)
    /// The "Finish this story" menu action -- see protocol.py's
    /// ConcludeStory. Carries a turn_id like speechStart/interrupt: it
    /// results in one more real response_text/turn_end pair the client
    /// must be able to attribute to a turn.
    case concludeStory(turnId: Int)
    /// Parent-adjustable story-length settings from the Settings screen --
    /// see protocol.py's UpdateSettings. Sent once after connecting and
    /// again whenever changed while connected.
    case updateSettings(targetTurns: Int, pageCount: Int)

    public func encode() -> String {
        // Field order and separators are fixed here (no JSONEncoder) so the
        // wire bytes are exact and predictable.
        switch self {
        case .speechStart(let turnId):
            return #"{"type":"speech_start","turn_id":\#(turnId)}"#
        case .speechEnd:
            return #"{"type":"speech_end"}"#
        case .interrupt(let turnId):
            return #"{"type":"interrupt","turn_id":\#(turnId)}"#
        case .objectSeen(let label):
            return #"{"type":"object_seen","label":"\#(Self.jsonEscaped(label))"}"#
        case .newStory:
            return #"{"type":"new_story"}"#
        case .syncDemoStories(let stories):
            let storiesJSON: [[String: Any]] = stories.map { story in
                [
                    "id": story.id,
                    "created_at": story.createdAt,
                    "turns": story.turns.map {
                        ["speaker": $0.speaker, "text": $0.text, "interrupted": $0.interrupted]
                    },
                    "shared_facts": story.sharedFacts,
                ]
            }
            let payload: [String: Any] = ["type": "sync_demo_stories", "stories": storiesJSON]
            guard let data = try? JSONSerialization.data(withJSONObject: payload),
                  let json = String(data: data, encoding: .utf8) else {
                return #"{"type":"sync_demo_stories","stories":[]}"#
            }
            return json
        case .listStories:
            return #"{"type":"list_stories"}"#
        case .getStory(let storyId):
            return #"{"type":"get_story","story_id":"\#(Self.jsonEscaped(storyId))"}"#
        case .concludeStory(let turnId):
            return #"{"type":"conclude_story","turn_id":\#(turnId)}"#
        case .updateSettings(let targetTurns, let pageCount):
            return #"{"type":"update_settings","target_turns":\#(targetTurns),"page_count":\#(pageCount)}"#
        }
    }

    /// Escapes the characters that could appear in a Vision classification
    /// label and break JSON's string syntax. Not a general-purpose JSON
    /// escaper -- control characters (U+0000-U+001F) are not handled,
    /// because the only value reaching this function is an identifier
    /// from Vision's fixed taxonomy.
    private static func jsonEscaped(_ s: String) -> String {
        var result = ""
        for scalar in s.unicodeScalars {
            switch scalar {
            case "\"": result += "\\\""
            case "\\": result += "\\\\"
            default: result.unicodeScalars.append(scalar)
            }
        }
        return result
    }
}

public enum ServerEvent: Sendable, Equatable {
    case transcriptPartial(String, turnId: Int)
    case transcriptFinal(String, turnId: Int)
    case responseText(String, turnId: Int)
    case turnEnd(turnId: Int)
    case error(String, turnId: Int)
    /// A story just concluded and its background storybook rewrite has
    /// started -- see server/tinytalk/protocol.py's encode_rewriting_started().
    /// Carries no turn_id or story_id: see SessionCoordinator's
    /// readyToShowTheEnd doc comment for why arrival of this event alone
    /// is NOT sufficient to know the concluding turn's audio has finished
    /// playing.
    case rewritingStarted
    /// The background rewrite finished (successfully or not) -- see
    /// encode_rewriting_done(). Carries no story_id; the client already
    /// knows which story it's waiting on (the most recent one) from
    /// rewritingStarted.
    case rewritingDone
    case storyList([SavedStorySummary])
    case storyDetail(SavedStoryDetail)
}

public enum ProtocolError: Error, Equatable {
    case malformed(String)
}

public func decodeServerEvent(_ raw: String) throws -> ServerEvent {
    guard let data = raw.data(using: .utf8),
          let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any] else {
        throw ProtocolError.malformed("control frame is not a valid JSON object: \(raw)")
    }
    guard let type = json["type"] as? String else {
        throw ProtocolError.malformed("control frame has no \"type\" field: \(raw)")
    }
    let turnId = json["turn_id"] as? Int ?? 0
    switch type {
    case "transcript_partial":
        return .transcriptPartial(json["text"] as? String ?? "", turnId: turnId)
    case "transcript_final":
        return .transcriptFinal(json["text"] as? String ?? "", turnId: turnId)
    case "response_text":
        return .responseText(json["text"] as? String ?? "", turnId: turnId)
    case "turn_end":
        return .turnEnd(turnId: turnId)
    case "error":
        return .error(json["message"] as? String ?? "", turnId: turnId)
    case "rewriting_started":
        return .rewritingStarted
    case "rewriting_done":
        return .rewritingDone
    case "story_list":
        let rawStories = json["stories"] as? [[String: Any]] ?? []
        return .storyList(rawStories.map(decodeStorySummary))
    case "story_detail":
        guard let id = json["id"] as? String else {
            throw ProtocolError.malformed("story_detail missing id: \(raw)")
        }
        return .storyDetail(decodeStoryDetail(json, id: id))
    default:
        throw ProtocolError.malformed("unknown server message type: \(type)")
    }
}

/// One entry of a story_list message -- see story_store.py's
/// list_stories(). Lenient about individual fields (matching this
/// file's existing style, e.g. turnId's default-to-0 above): a
/// malformed summary should degrade gracefully, not take down the
/// whole list.
private func decodeStorySummary(_ json: [String: Any]) -> SavedStorySummary {
    SavedStorySummary(
        id: json["id"] as? String ?? "",
        title: json["title"] as? String,
        createdAt: parseISODate(json["created_at"] as? String) ?? Date(),
        pageCount: json["page_count"] as? Int ?? 0,
        rewriteStatus: RewriteStatus(rawValue: json["rewrite_status"] as? String ?? "") ?? .pending
    )
}

/// story_detail's payload -- see session.py's handle_get_story(). `id`
/// is already validated present by the caller (decodeServerEvent);
/// everything else defaults leniently, same reasoning as
/// decodeStorySummary above.
private func decodeStoryDetail(_ json: [String: Any], id: String) -> SavedStoryDetail {
    let rawPages = json["pages"] as? [[String: Any]] ?? []
    let pages = rawPages.map { StoryPage(text: $0["text"] as? String ?? "") }
    return SavedStoryDetail(
        id: id,
        title: json["title"] as? String,
        pages: pages,
        epilogue: json["epilogue"] as? String,
        rewriteStatus: RewriteStatus(rawValue: json["rewrite_status"] as? String ?? "") ?? .pending
    )
}

/// Parses story_store.py's `datetime.now(timezone.utc).isoformat()`
/// output (e.g. "2026-09-09T12:00:00.123456+00:00") -- includes
/// microsecond fractional seconds, which ISO8601DateFormatter only
/// parses with .withFractionalSeconds explicitly set. Falls back to a
/// formatter without that option for a timestamp with no fractional part
/// at all (e.g. the exact strings used in this file's own tests).
private func parseISODate(_ raw: String?) -> Date? {
    guard let raw else { return nil }
    let withFractional = ISO8601DateFormatter()
    withFractional.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
    if let date = withFractional.date(from: raw) { return date }
    let withoutFractional = ISO8601DateFormatter()
    withoutFractional.formatOptions = [.withInternetDateTime]
    return withoutFractional.date(from: raw)
}
